### PR TITLE
Add implementation for pagination and search on group membership

### PR DESCRIPTION
**Pull request by volsusdev**
Hi Keycloak team,
we have extended REST API to support pagination and search on group membership endpoint of a user. In addition to this Angular frontend was adapted to use this features to provide a paginated and searchable list for available groups and membership.
FYI: Datatype of membership results was changed from Set to List since we need an order of groups to keep pages consistent.

Best regards
Volker

### CODE REVIEW

**Comment by volsus83** [comment]
We've also created a ticket for this pull request: https://issues.jboss.org/browse/KEYCLOAK-8613

**Review by hmlnarik — COMMENTED** [review]
Thanks for the contribution. Few general comments (after I've checked the code, there are repeating patterns):

- Don't just change `Set` to `List`. It negatively impacts performance of other parts of the code. Prefer returning `LinkedHashSet` when needed.
- We cannot change `UserModel`. This is a supported SPI and has to stay as it is.
- The order / comparator needs to be configurable to allow also other fields to be used for ordering.
- Do not use star imports, revert the already changed back.

File: themes/src/main/resources/theme/base/admin/resources/js/app.js [hunk 1]
```diff
@@ -1419,19 +1413,7 @@ module.config([ '$routeProvider', function($routeProvider) {
             },
             controller : 'ClientDetailCtrl'
         })
-        .when('/create/client-scope/step-1/:realm', {
```
Comment (mposolda) [inline themes/src/main/resources/theme/base/admin/resources/js/app.js]: It seems you did not correctly rebased with the latest master? You seem to remove some existing functionality, which was added in the meantime.

File: themes/src/main/resources/theme/base/admin/resources/js/services.js [hunk 2]
```diff
@@ -349,18 +349,8 @@ module.factory('RegisterRequiredAction', function($resource) {
     });
 });
 
-module.factory('RealmLDAPConnectionTester', function($resource, $httpParamSerializer) {
```
Comment (mposolda) [inline themes/src/main/resources/theme/base/admin/resources/js/services.js]: Again, you seem to break things when rebasing with latest master. You removed recently added changes and replaced with your local ones.

File: themes/src/main/resources/theme/base/admin/resources/js/services.js [hunk 3]
```diff
@@ -1404,29 +1387,23 @@ module.factory('TimeUnit2', function() {
     var t = {};
 
     t.asUnit = function(time) {
-
         var unit = 'Minutes';
-
         if (time) {
-            if (time == -1) {
```
Comment (mposolda) [inline themes/src/main/resources/theme/base/admin/resources/js/services.js]: Not 100% sure, but looks like those changes in the time is also something, which should be removed from your PR (likely changed by accident when rebasing)

**Review by mposolda — COMMENTED** [review]
@volsus83 Thanks for the PR and sorry it took longer for review :( It looks like nice contribution, however there will be some additional work needed IMO.

+1 to Hynek's comments.

@hmlnarik @stianst Regarding UserModel changes, question is how to proceed as we should be able to add new functionality to UserModel? The easiest would be to add default implementations to the UserModel interface and those default implementations will just delegate to "getGroups()" and then programatically filter (or count) groups to be returned? We can also add note to the migration guide that customers are encouraged to implement new methods more properly/effectively in their providers if they want to profit from the performance improvements of groups pagination

- The PR probably breaks LDAP when HardcodedLDAPStorageMapper or GroupLDAPStorageMapper are used. The PR should be updated to correctly handle groups in case of LDAP integration enabled. The proper implementation would be non-trivial as it will need to "merge" the group memberships from both LDAP and DB and ensure they work correctly regarding pagination etc. I understand that if you're not interested in LDAP, you likely don't want to do it. But at least minimal update needed would be to:
-- ensure that UserModel delegate methods delegate to "getGroups()" (if we agree to add default methods on USerModel, then no changes will be defacto needed in LDAP mappers)
-- Add the LDAP test, which will call paginated "getGroups" and will check that it works correctly. Will be good if you can use admin client and new methods you added on the UserResource interface

- GroupTest and UserStorageTest should be also improved/extended to add some test for the paginated methods


- I think you broke some admin console functionality due the incorrect rebase of the angular files. Some more comments inline. Did you try to run admin console tests? BTV I think that some update in admin console tests will be needed as they currently don't use pagination for groups. See how to run admin console tests on your laptop - https://github.com/keycloak/keycloak/blob/master/testsuite/integration-arquillian/HOW-TO-RUN.md#admin-console-ui-tests . I suggest to try at least one browser.

**Comment by unly** [comment]
Thank you @mposolda and @hmlnarik for your review.
Since my colleague @volsus83 is currently out of office, I would have a look at his code and your comments and try to adapt it accordingly.

**Comment by unly** [comment]
As far as I see is the sorting in the `getGroups` function tied to the order by statement in the `createGetGroupsQuery` function which makes it harder to change it dynamically. Sorting in the application would require to load all groups into memory which will result in a performance issue. Especially since the group query is based on a n+1 query. In our deployment we have users with several thousand group memberships and this wouldn't solve any problems. Dropping the sorting would speed up the sql query as needed but at a price.

I understand the impact of changing the `UserModel`. The thing is that we need somehow to access the database from the REST entry point in `UserResource`. And the right place to do so would be the `UserAdapter` in my opinion. Do you agree?
Maybe we can have an interface extending the existing `UserModel` to avoid having changes in unnecessary classes.

File: model/infinispan/src/main/java/org/keycloak/models/cache/infinispan/UserAdapter.java [hunk 4]
```diff
@@ -357,9 +349,41 @@ public void deleteRoleMapping(RoleModel role) {
...
 
         }
+        groups.sort(Comparator.comparing(GroupModel::getName));
         return groups;
     }
 
+    @Override
+    public List<GroupModel> getGroups(Integer first, Integer max) {
+        return getGroups().stream()
+                .sorted(Comparator.comparing(GroupModel::getName))
```
[the remark this answers is not in the document]
Comment (mposolda) [reply model/infinispan/src/main/java/org/keycloak/models/cache/infinispan/UserAdapter.java]: On the other hand, if you know that model returned groups sorted by name, there is no need to sort them again in ModelToRepresentation? Can this additional sorting be removed from the ModelToRepresentation methods?

**Comment by unly** [comment]
@mposolda I updated the pull request. Not sure though what to add on tests for the LDAP integration.

**Comment by mposolda** [comment]
@unly Thanks a lot for the changes!

I see there are still the changes in the users.js and services.js related to RealmLDAPConnectionTester and TimeUnits. Those changes should be removed from this PR. It looks that those changes were added by accident (looks that you or @volsus83 incorrectly rebased with latest master), not intentionally as those are unrelated to groups pagination? See inline comments for more details.

Regarding LDAP, you can take a look at LDAPGroupMapperTest. We need some new test method here, which will use those new methods on UserModel (paginated getGroups etc) and test that correct results are returned. Is it more clear now?

**Comment by unly** [comment]
@mposolda I did the rebase again and tried to minimize the changes to avoid unnecessary conflicts. There were lots of changes done by my colleague in the JavaScript files but I'm not that deep into the code. As far as I can tell the functionality was the same in my tests. Also I added some test to the `LDAPGroupMapperTest.java`. Can you have a look? Thanks

**Comment by mposolda** [comment]
@unly Thanks! I've added last 2 comments with the text "This change should be removed from the PR", which seem to be the last things, which were added by accident to this PR and should be removed from it.

Once you do it, we may need to run the DB matrix tests against this PR to check it didn't break any DB. I will ping you if there is any issue with it.

**Comment by unly** [comment]
The accidentally added lines should be removed by now. As far as I understand the rest of the code is required to have the pagination in the UI. If there is something left can you please add a comment to that line? The ones you previously mentioned are gone.

**Comment by mposolda** [comment]
@unly Sorry, I've added comments to those 2 lines yesterday, but those inline comments weren't pushed for some reason. I've updated it now - there are still 2 calls to RealmLDAPConnectionTester, which should be removed from the PR

**Review by hmlnarik — COMMENTED** [review]
Thank you @volsus83 and @unly. Java part lgtm. I've checked the admin console manually and it works; I've not gone deep into JS code though.

**Comment by mposolda** [comment]
@hmlnarik Thanks for the review! It looks good to me as well.

@vramik Is it possible to run DB matrix against this PR? Thanks!

**Comment by mposolda** [comment]
@keycloak-ci-bot test

**Comment by vramik** [comment]
@mposolda I've run the DB matrix and there are results:
tests failed with 

- MSSQL2016, I've commented https://issues.jboss.org/browse/KEYCLOAK-8613 with more info.
- mysql57 and Oracle12gRAC are currently broken in our internal instances so it cannot be easily tested atm
- with mariadb failed one test: org.keycloak.testsuite.federation.ldap.LDAPGroupMapperTest#test02_readOnlyGroupMappings(LDAPGroupMapperTest.java:281) AssertionError: expected:<2> but was:<3>


postgres passed.

**Comment by mposolda** [comment]
@vramik Thanks!

@unly Both MariaDB and MSSQL have docker images publicly available (although for MSSQL you need to approve the licence). Do you have possibility to try them and simulate issue with them?

**Comment by unly** [comment]
I started having a look at this failed test and the problem is the SQL query to the database. 

>     * with mariadb failed one test: org.keycloak.testsuite.federation.ldap.LDAPGroupMapperTest#test02_readOnlyGroupMappings(LDAPGroupMapperTest.java:281) AssertionError: expected:<2> but was:<3>

The search on cached groups uses a case insensitive comparison

````
default Set<GroupModel> getGroups(String search, int first, int max) {
    return getGroups().stream()
        .filter(group -> search == null || group.getName().toLowerCase().contains(search.toLowerCase()))
        .skip(first)
        .limit(max)
        .collect(Collectors.toCollection(LinkedHashSet::new));
}
````

and the query uses:
````
if (Objects.nonNull(search) && !search.isEmpty()) {
    predicates.add(builder.like(join.get("name"), builder.literal("%" + search + "%")));
}
````

Changing the query to case insensitive comparison returns 3 groups for that specific test case.
I tried to run the database test and failed. Commented on the jira issue.

**Comment by mposolda** [comment]
@vramik Cool, Thanks!

**Comment by mposolda** [comment]
@keycloak-ci-bot test

**Review by mposolda — APPROVED** [review]

**Pull request merged**